sw/applications/example_spi_read/main.c

```diff
@@ -0,0 +1,180 @@
+    spi.base_addr = mmio_region_from_addr((uintptr_t)SPI_HOST_START_ADDRESS);
+    #else
+    spi.base_addr = mmio_region_from_addr((uintptr_t)SPI_FLASH_START_ADDRESS);
+    #endif
+
+    // Init SPI host and SPI<->Flash bridge parameters 
+    global_status = w25q128jw_init(spi);
+    if (global_status != FLASH_OK) return EXIT_FAILURE;
+
+    // Test simple read
+    PRINTF("Testing simple read...\n\r");
+    global_status = test_read(TEST_BUFFER, LENGTH);
+    // if (global_status != FLASH_OK) return EXIT_FAILURE;
+
+    PRINTF("Testing simple read with DMA...\n\r");
+    // Test simple read with DMA
+    global_status = test_read_dma(TEST_BUFFER, LENGTH);
+    // if (global_status != FLASH_OK) return EXIT_FAILURE;
+
+    PRINTF("Testing quad read...\n\r");
+    // Test quad read
+    global_status = test_read_quad(TEST_BUFFER, LENGTH);
+    // if (global_status != FLASH_OK) return EXIT_FAILURE;
+
+    PRINTF("Testing quad read with DMA...\n\r");
+    // Test quad read with DMA
+    global_status = test_read_quad_dma(TEST_BUFFER, LENGTH);
+    // if (global_status != FLASH_OK) return EXIT_FAILURE;
+
+    PRINTF("\n--------TEST FINISHED--------\n");
+    if (global_errors == 0) {
+        PRINTF("All tests passed!\n\r");
+        return EXIT_SUCCESS;
+    } else {
+        PRINTF("Some tests failed!\n\r");
+        return EXIT_FAILURE;
+    }
+    
+}
+
+w25q_error_codes_t test_read(uint32_t *test_buffer, uint32_t len) {
+    // Read from flash memory at the same address
+    global_status = w25q128jw_read(test_buffer, flash_data, len);
+    if (global_status != FLASH_OK) return FLASH_ERROR;
+
+    // Check if what we read is correct (i.e. flash_data == test_buffer)
+    check_result(test_buffer, len);
+
+    return FLASH_OK;
+}
+
+w25q_error_codes_t test_read_dma(uint32_t *test_buffer, uint32_t len) {
+    // Read from flash memory at the same address
+    global_status = w25q128jw_read(test_buffer, flash_data, len);
+    if (global_status != FLASH_OK) return FLASH_ERROR;
+
+    // Check if what we read is correct (i.e. flash_data == test_buffer)
+    check_result(test_buffer, len);
+
+    return FLASH_OK;
+}
+
+w25q_error_codes_t test_read_quad(uint32_t *test_buffer, uint32_t len) {
+    // Read from flash memory at the same address
+    global_status = w25q128jw_read(test_buffer, flash_data, len);
+    if (global_status != FLASH_OK) return FLASH_ERROR;
+
+    // Check if what we read is correct (i.e. flash_data == test_buffer)
+    check_result(test_buffer, len);
+
+    return FLASH_OK;
+}
+
+w25q_error_codes_t test_read_quad_dma(uint32_t *test_buffer, uint32_t len) {
+    // Read from flash memory at the same address
+    global_status = w25q128jw_read(test_buffer, flash_data, len);
+    if (global_status != FLASH_OK) return FLASH_ERROR;
+
+    // Check if what we read is correct (i.e. flash_data == test_buffer)
+    check_result(test_buffer, len);
+
+    return FLASH_OK;
+}
+
+void check_result(uint32_t *test_buffer, uint32_t len) {
```

Comment:
check_result should return the number of errors

Reply:
Ok, I was keeping track of the errors in a global variable. I will now change it to a global-variable safe implementation.